audio/mods/mod_xm_s3m.cpp

```diff
@@ -0,0 +1,1353 @@
+	memset(mixBuf, 0, (tickLen + 65) * 4 * sizeof(int));
+	/* Resample. */
+	for (int idx = 0; idx < _module.numChannels; idx++) {
+		Channel &channel = _channels[idx];
+		resample(channel, mixBuf, 0, (tickLen + 65) * 2, _sampleRate * 2);
+		updateSampleIdx(channel, tickLen * 2, _sampleRate * 2);
+	}
+	downsample(mixBuf, tickLen + 64);
+	volumeRamp(mixBuf, tickLen);
+	tick();
+	return tickLen;
+}
+
+void ModXmS3mConvertor::xmToWav(byte *wavData, int duration) {
+	// write header and sound informations
+	int length = duration * 4 + 40;
+
+#if DEBUG_DUMP_FILE
+	Common::DumpFile *writeStream = new Common::DumpFile();
+	Common::String songname = Common::String::format("%s.wav", _module.name);
+	writeStream->open(songname);
+#else
+	Common::MemoryWriteStream *writeStream = new Common::MemoryWriteStream(wavData, length);
+#endif
+	char sig1[4] = { 'R', 'I', 'F', 'F' };
+	writeStream->write(sig1, 4);
+	writeStream->writeUint32LE(duration * 4 + 36);
+	char sig2[8] = { 'W', 'A', 'V', 'E', 'f', 'm', 't', ' ' };
+	writeStream->write(sig2, 8);
+	writeStream->writeUint32LE(16);
+	writeStream->writeUint32LE(0x00020001);
+	writeStream->writeUint32LE(_sampleRate);
+	writeStream->writeUint32LE(_sampleRate * 4);
+	writeStream->writeUint32LE(0x00100004);
+	char sig3[4] = { 'd', 'a', 't', 'a' };
+	writeStream->write(sig3, 4);
+	writeStream->writeUint32LE(length);
+
+	seek(0);
+
+	// write data
+	int mixBuf[16384] = { 0 };
+	while (writeStream->pos() < length) {
+		int samples = getAudio(mixBuf) * 2;
+		for (int idx = 0; idx < samples; ++idx) {
+			int ampl = mixBuf[idx];
+			if (ampl > 32767) {
+				ampl = 32767;
+			}
+			if (ampl < -32768) {
+				ampl = -32768;
+			}
+			writeStream->writeByte(ampl & 0xFF);
+			writeStream->writeByte((ampl >> 8) & 0xFF);
+		}
+	}
+	writeStream->finalize();
+#if DEBUG_DUMP_FILE
+	writeStream->close();
+#endif
+	delete writeStream;
+}
+
+void ModXmS3mConvertor::setSequencePos(int pos) {
+	if (pos >= _module.sequenceLen) {
+		pos = 0;
+	}
+	_breakPos = pos;
+	_nextRow = 0;
+	_tick = 1;
+	_globalVol = _module.defaultGvol;
+	_speed = _module.defaultSpeed > 0 ? _module.defaultSpeed : 6;
+	_tempo = _module.defaultTempo > 0 ? _module.defaultTempo : 125;
+	_plCount = _plChan = -1;
+
+	// play count
+	if (_playCount) {
+		delete[] _playCount[0];
+		delete[] _playCount;
+	}
+	_playCount = new int8 *[_module.sequenceLen];
+	memset(_playCount, 0, _module.sequenceLen * sizeof(int8 *));
+	int len = initPlayCount(_playCount);
+	_playCount[0] = new int8[len];
+	memset(_playCount[0], 0, len * sizeof(int8));
+	initPlayCount(_playCount);
+
+	for (int idx = 0; idx < _module.numChannels; ++idx) {
+		initChannel(idx);
+	}
+	memset(_rampBuf, 0, 128 * sizeof(int));
+	tick();
+}
+
+} // End of namespace Modules
+
+namespace Audio {
+
+AudioStream *makeModXmS3mStream(Common::SeekableReadStream *stream, int rate, int interpolation) {
+	Modules::ModXmS3mConvertor *modXmS3mConvertor = new Modules::ModXmS3mConvertor(stream, rate, interpolation);
```

Comment:
`modXmS3mConvertor` never freed, resulting in a memory leak.

Reply:
True, I'll correct it.